File: 백준/Silver/13702. 이상한 술집/이상한 술집.py
import sys

data = sys.stdin.read().strip().split('\n')
N, K = map(int, data.pop(0).split())
amounts = list(map(int, data))


left = 1
right = max(amounts) * 2
answer = 0

while left <= right:
    mid = (left + right) // 2
    count = 0
    for amount in amounts:
        count += amount // mid  

    if count >= K:
        answer = mid
        left = mid + 1
    else:
        right = mid - 1

print(answer)

File: 백준/Silver/13702. 이상한 술집/이상한 술집.js
const input = require('fs')
    .readFileSync(0, 'utf-8')
    .trim()
    .split('\n');

const [N, K] = input.shift().split(' ').map(Number);
const amounts = input.map(Number);

let left = 1;
let right = Math.max(...amounts);
let answer = 0;

while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    let count = 0;

    for (const amount of amounts) {
        count += Math.floor(amount / mid);
    }

    if (count >= K) {
        answer = mid;
        left = mid + 1;
    } else {
        right = mid - 1;
    }
}

console.log(answer);